src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject_IIntellisenseBuildTarget.cs

```diff
@@ -0,0 +1,129 @@
+
+            UpdateHostDiagnostics(succeed, reason);
+
+            if (_pushingChangesToWorkspaceHosts)
+            {
+                // set workspace reference info
+                ProjectTracker.NotifyWorkspaceHosts(host => host.OnIntellisenseBuildResult(Id, succeed));
+            }
+        }
+
+        private void UpdateHostDiagnostics(bool succeed, string reason)
+        {
+            if (!succeed)
+            {
+                // report intellisense build failure to error list
+                this.HostDiagnosticUpdateSource?.UpdateDiagnosticsForProject(
+                    _id, s_diagnosticKey, SpecializedCollections.SingletonEnumerable(CreateIntellisenseBuildFailureDiagnostic(reason)));
+            }
+            else
+            {
+                // clear intellisense build failure diagnostic from error list.
+                this.HostDiagnosticUpdateSource?.ClearDiagnosticsForProject(_id, s_diagnosticKey);
+            }
+        }
+
+        private DiagnosticData CreateIntellisenseBuildFailureDiagnostic(string reason)
+        {
+            // log intellisense build failure
+            Logger.Log(FunctionId.IntellisenseBuild_Failed, KeyValueLogMessage.Create(m => m["Reason"] = reason ?? string.Empty));
+
+            var isEnabledByDefault = true;
+            var warningLevel = 0;
+
+            return new DiagnosticData(
+                IDEDiagnosticIds.IntellisenseBuildFailedDiagnosticId,
+                FeaturesResources.ErrorCategory,
+                ServicesVSResources.IntellisenseBuildFailedMessage,
+                ServicesVSResources.ResourceManager.GetString(nameof(ServicesVSResources.IntellisenseBuildFailedMessage), CodeAnalysis.Diagnostics.Extensions.s_USCultureInfo),
+                DiagnosticSeverity.Warning,
+                isEnabledByDefault, warningLevel, Workspace, Id,
+                title: ServicesVSResources.IntellisenseBuildFailedTitle,
+                description: GetDescription(reason));
+        }
+
+        private string GetDescription(string reason)
+        {
+            var logFile = string.Format(ServicesVSResources.IntellisenseBuildFailedDescription, GetLogFile());
+            if (string.IsNullOrWhiteSpace(reason))
+            {
+                return logFile;
+            }
+
+            return string.Join(Environment.NewLine, logFile, string.Empty, ServicesVSResources.IntellisenseBuildFailedDescriptionExtra, reason);
+        }
+
+        private string GetLogFile()
+        {
+            var manager = BuildManager.DefaultBuildManager;
+            if (manager == null)
+            {
+                return GeneralLogFile;
+            }
+
+            var buildParameters = s_fieldInfo.GetValue(manager) as BuildParameters;
+            if (buildParameters == null)
+            {
+                return GeneralLogFile;
+            }
+
+            var fileLogger = buildParameters.Loggers?.OfType<FileLogger>()?.FirstOrDefault();
```

Comment:
I tested it with Roslyn. it works because when this is called, things run on UI thread and everything is serialized.